app/services/promo/registration_stats_report_generator.rb

```diff
@@ -2,113 +2,140 @@
-      events = JSON.parse(promo_registration.stats)
-
-      # Select only those within date range supplied
-      events_within_report_period = events.select { |event|
-        (event["ymd"].to_date >= @start_date) && (event["ymd"].to_date <= @end_date)
-      }
-
-      if @reporting_interval == PromoRegistration::RUNNING_TOTAL
-        base_case = {PromoRegistration::RETRIEVALS => 0, PromoRegistration::FIRST_RUNS => 0, PromoRegistration::FINALIZED => 0 }
-        running_total_report_contents = events_within_report_period.reduce(base_case) { |aggregate_stats, event|
-          aggregate_stats[PromoRegistration::RETRIEVALS] += event[PromoRegistration::RETRIEVALS]
-          aggregate_stats[PromoRegistration::FIRST_RUNS] += event[PromoRegistration::FIRST_RUNS]
-          aggregate_stats[PromoRegistration::FINALIZED] += event[PromoRegistration::FINALIZED]
-          aggregate_stats.slice(PromoRegistration::RETRIEVALS, PromoRegistration::FIRST_RUNS, PromoRegistration::FINALIZED)
-        }
-
-        report_contents_for_referral_code = {
-          @start_date => running_total_report_contents
-        }
-
-        report_contents["#{promo_registration.referral_code}"] = report_contents_for_referral_code
-      else
-        report_contents_for_referral_code = empty_report_contents_for_referral_code(@start_date, @end_date, @reporting_interval)
-        events_within_report_period.each do |event|
-          interval_start_date = coerce_date_to_start_or_end_of_reporting_interval(event["ymd"].to_date, @reporting_interval, true)
-          existing_event = report_contents_for_referral_code[interval_start_date]
-          existing_event[PromoRegistration::RETRIEVALS] += event[PromoRegistration::RETRIEVALS]
-          existing_event[PromoRegistration::FIRST_RUNS] += event[PromoRegistration::FIRST_RUNS]
-          existing_event[PromoRegistration::FINALIZED] += event[PromoRegistration::FINALIZED]
+
+    events = fetch_stats
+    events = select_events_within_report_range(events)
+    events = fill_in_events_with_no_activity(@referral_codes, events)
+
+    events_by_referral_codes = group_events_by_referral_code(events)
+
+    events_by_referral_codes.each do |referral_code, events|
+      events = order_events_by_date(events)
+      events_by_referral_code_for_interval = group_events_by_date(events)
+
+      # Sum the events per interval, or per interval per country if @geo
+      per_interval_totals = {}
+      events_by_referral_code_for_interval.each do |date, events|
+        if @geo
+          events_by_referral_code_for_interval_by_country = group_events_by_country(events)
+
+          per_country_totals = {}
+          events_by_referral_code_for_interval_by_country.each do |country, events|
+            per_country_totals[country] = sum_totals(events)
+          end
+
+          # If there are any events for any country other than N/A, remove the blank country data
+          per_country_totals.delete("N/A") if per_country_totals.keys.count > 1
```

Comment:
Where is this `N/A` coming from?

Reply:
If there were no events for a given referral code on a specific day, then there will be no recording of that day in the /geoStatsByReferralCode endpoint.  However it's a requirement that we explicitly show 0 when there were no events on that day.

Without country data it was easy enough to just to create 0 values for all possible days, then insert the real data from the referral server:  If there was a download on that day, then we can sum 0 with the number of downloads and get the correct amount.  If there wasn't then 0 was the correct thing to display

With country data it's a little trickier, since we need to associate a country with these 0 inserts.  In this case I'm inserting "N/A".  If there are downloads for that day and there is a country associated (say United States), then we want to remove it since it doesn't make sense to have an entry with N/A if there is another entry on the same day for United States.  Hope that makes sense.

I'm inserting in these blank events here:
https://github.com/brave-intl/publishers/pull/1462/files#diff-bc3a9bb0922dd200752c37c9a23e4df3R122

There is probably some room for improvement here, though a better solution that's simple has escaped me thus far.